Add new search filters

TobyRet opened the pull request:
### Context

```
As a... user searching for teacher training courses
I need to... filter results to only show those that match my requirements
So that... it is easier for to search and compare courses
```

### Changes proposed in this pull request
- Add new results page filters as per https://find-prototype.herokuapp.com/results.
- These filters are behind a feature flag (`:new_filters`)

### Guidance to review
- To see the new features, pull the branch and add `new_filters: true` to your `development.local.yaml`

```
feature_flags:
  cycle_ending_soon: false
  cycle_has_ended: false
  display_apply_button: true
  new_filters: true
```


- Please let me know if I have added any HTML, CSS or accessibility violations
- Note that the new filters are not a good candidate for govuk_form_builder, as validations are not required, plus there are some complications around hydrating filter form values from params
- Note that once the feature flag is removed, some refactoring will be required to remove old filters and specs.  
- There is some logic in the new filter views that would best be placed elsewhere (e.g helpers) - some of this may be turned into cards
- Commits will be tidied up before merging

<img width="564" alt="new_filters" src="https://user-images.githubusercontent.com/5256922/107493337-64268300-6b85-11eb-90f4-dda0feade3b1.png">

### Trello card
https://trello.com/c/A9f4PnzV/2931-%F0%9F%97%BA-dev-new-search-filters

### Checklist

- [ ] Rebased master
- [ ] Cleaned commit history
- [ ] Tested by running locally

TobyRet commented on app/helpers/result_filters/subject_helper.rb:
```diff
@@ -1,6 +1,7 @@
 module ResultFilters
   module SubjectHelper
     include FilterParameters
+    include CsharpRailsSubjectConversionHelper
```
@davidgisbey - I think we need a tech debt card to remove the csharp subjecct id conversion - I can't imagine anyone would still be querying the site with legacy ids ...

davidgisbey replied:
Great. I'll card it up 👍

paulrobertlloyd commented on app/views/new_filters/_all.html.erb:
```diff
@@ -0,0 +1,21 @@
+      <%= render 'new_filters/study_type_filter', form: form %>
+      <%= render 'new_filters/qualifications_filter', form: form %>
+      <%= render 'new_filters/salary_filter', form: form %>
+      <div class="govuk-form-group app-filter__group">
+        <%= form.submit "Apply filters", name: nil, class: "govuk-button", data: { qa: "apply-filters" } %>
+      </div>
```
```suggestion
      <%= form.submit "Apply filters", name: nil, class: "govuk-button govuk-!-margin-bottom-1", data: { qa: "apply-filters" } %>
```

paulrobertlloyd commented:
Sorry, this was a later change; salary should be a checkbox like the SEND and vacancy filter options:

| Before | After |
| - | - |
| <img width="289" alt="Screenshot 2021-02-10 at 10 59 17" src="https://user-images.githubusercontent.com/813383/107501236-14e55000-6b8f-11eb-88e4-4d121435d309.png"> | <img width="309" alt="Screenshot 2021-02-10 at 10 59 24" src="https://user-images.githubusercontent.com/813383/107501240-16167d00-6b8f-11eb-96b1-295c72c7e945.png"> |

paulrobertlloyd commented on app/views/new_filters/_area_provider_filter.html.erb:
```diff
@@ -0,0 +1,33 @@
+    </h3>
+  </legend>
+  <p class="govuk-body-s govuk-!-margin-bottom-2" data-qa="area_or_provider_name">
+    <!-- Move to helper -->
```
Use ERB comment to hide from frontend

```suggestion
    <%# Move to helper %>
```

paulrobertlloyd commented:
Clicking the label for SEND and vacancy filters doesn’t work. This is because the `for` attribute doesn’t match the `id` of the corresponding input:

* `subject_send` and `senCourses`
* `hasvacancies_true` and `hasvacancies`

paulrobertlloyd commented:
I’ve updated the design we use on mobile; rather than a toggle, changed it so that filters now appear as an overlay (see: https://bat-design-history.netlify.app/find-teacher-training/searching-by-location-iteration/#changing-how-search-filters-look-on-a-mobile-device)

I’d actually be tempted to make this change in another PR, though.

paulrobertlloyd commented on app/views/new_filters/_area_provider_filter.html.erb:
```diff
@@ -0,0 +1,33 @@
+      <% if request.parameters['query'] %>
+        Provider
+      <% else %>
+        Area
```
Until we’ve moved to the new location model, let’s stick to saying ‘location’ for now.

```suggestion
        Location
```

TobyRet replied:
👍

paulrobertlloyd commented on app/views/new_filters/_area_provider_filter.html.erb:
```diff
@@ -0,0 +1,33 @@
+  </p>
+
+  <% if request.parameters['l'] == '3' %>
+    <%= link_to "Change provider or choose a location",
+      @results_view.filter_path_with_unescaped_commas(location_path),
+      class: "govuk-link",
+      data:  { qa: "filters__area_and_provider_link" } %>
```
```suggestion
    <%= govuk_link_to "Change provider or choose a location",
      @results_view.filter_path_with_unescaped_commas(location_path),
      data:  { qa: "filters__area_and_provider_link" } %>
```

TobyRet replied:
👍

paulrobertlloyd commented on app/views/new_filters/_area_provider_filter.html.erb:
```diff
@@ -0,0 +1,33 @@
+      class: "govuk-link",
+      data:  { qa: "filters__area_and_provider_link" } %>
+  <% else %>
+    <%= link_to "Change location or choose a provider",
+      @results_view.filter_path_with_unescaped_commas(location_path),
+      class: "govuk-link",
+      data:  { qa: "filters__area_and_provider_link" } %>
```
```suggestion
    <%= govuk_link_to "Change location or choose a provider",
      @results_view.filter_path_with_unescaped_commas(location_path),
      data:  { qa: "filters__area_and_provider_link" } %>
```

TobyRet replied:
👍

paulrobertlloyd commented on app/views/new_filters/_study_type_filter.html.erb:
```diff
@@ -0,0 +1,36 @@
+    <h3 class="govuk-fieldset__heading">Study type</h3>
+  </legend>
+  <div class="govuk-checkboxes govuk-checkboxes--small">
+    <!-- Move this var to a method in a helper -->
```
```suggestion
    <%# Move this var to a method in a helper %>
```

TobyRet replied:
👍

paulrobertlloyd commented on app/views/new_filters/_subjects_filter.html.erb:
```diff
@@ -0,0 +1,15 @@
+      <%= form.hidden_field(:subjects, multiple: true, value: csharp_subject_id) %>
+    <% end %>
+  <% end %>
+  <%= link_to "Change",
+    @results_view.filter_path_with_unescaped_commas(subject_path),
+    class: "govuk-link",
+    data:  { qa: "link" } %>
```
```suggestion
  <%= govuk_link_to "Change",
    @results_view.filter_path_with_unescaped_commas(subject_path),
    data:  { qa: "link" } %>
```

TobyRet replied:
👍

paulrobertlloyd commented on app/webpacker/styles/components/_filter.scss:
```diff
@@ -0,0 +1,53 @@
+    &:last-child {
+      margin-bottom: 0;
+    }
+
+    //.govuk-fieldset__legend {
+    //  @include govuk-font(16);
+    //  color: $govuk-secondary-text-colour;
+    //}
```
```suggestion
```

TobyRet replied:
👍

paulrobertlloyd reviewed (changes requested):
Comments inline.

TobyRet commented:
> I’ve updated the design we use on mobile; rather than a toggle, changed it so that filters now appear as an overlay (see: https://bat-design-history.netlify.app/find-teacher-training/searching-by-location-iteration/#changing-how-search-filters-look-on-a-mobile-device)
> 
> I’d actually be tempted to make this change in another PR, though.

Yep I can create another card for this

> Clicking the label for SEND and vacancy filters doesn’t work. This is because the `for` attribute doesn’t match the `id` of the corresponding input:
> 
> * `subject_send` and `senCourses`
> * `hasvacancies_true` and `hasvacancies`

Good spot. That's fixed now

davidgisbey commented on app/views/new_filters/_location_provider_filter.html.erb:
```diff
@@ -0,0 +1,30 @@
+      <% if @filters_view.provider_query? %>
+        Provider
+      <% else %>
+        Location
```
This says 'Area' in the prototype. I think location might make more sense to use for now though. Happy to defer to Paul though.

paulrobertlloyd replied:
Yep, we’ll use ‘Location’ for now, can change it to ‘Area’ when we’re using the new location… area model.

davidgisbey replied:
Should have checked your comments 🤦

davidgisbey commented on app/views/new_filters/_location_provider_filter.html.erb:
```diff
@@ -0,0 +1,30 @@
+    <%= govuk_link_to "Change location or choose a provider",
+      @results_view.filter_path_with_unescaped_commas(location_path),
+      data:  { qa: "filters__area_and_provider_link" } %>
+  <% end %>
```
Just had a play with the filters. This is working nicely when you first hit the results page, but if you use the filters it doesn't retain the selected providers.

TobyRet replied:
Yep I see what the issue is here ... we need to pass in the location query params as hidden fields. Have pushed a fix

davidgisbey replied:
Do we also need to tdo his for the :query param to retain the selected provider?

davidgisbey commented on app/view_objects/results_view.rb:
```diff
@@ -205,6 +205,14 @@ def subjects
+    @all_subjects ||= Subject.select(:subject_name, :subject_code).order(:subject_name).all
+  end
+
+  def all_subjects_no_ordered
```
`all_subjects_not_ordered`

davidgisbey commented on app/views/results/index.html.erb:
```diff
@@ -1,6 +1,11 @@
+<% if FeatureFlag.active?(:new_filters) %>
+  <h1 class="govuk-heading-l" data-qa="heading">
+    <span class="govuk-caption-l">Teacher training courses</span>
+    <%= "#{@results_view.number_of_courses_string} found" %>
```
just making a note that this has been removed? I might find out why in a later commit 👍

TobyRet replied:
Sorry that was an error. Fixed now

davidgisbey commented on spec/features/new_results_page_filters/subjects_spec.rb:
```diff
@@ -0,0 +1,329 @@
+      end
+    end
+
+    context 'with subjects selected' do
```
I might be missing something, but this and the above test seem to be testing the same thing.

TobyRet replied:
Yes agree! Have removed the dupped spec

davidgisbey commented on spec/features/new_results_page_filters/subjects_spec.rb:
```diff
@@ -0,0 +1,329 @@
+    end
+  end
+
+  context 'with only SEND courses selected' do
```
Again i might be missing something, but I'm not 100% sure why we're testing this. Is SEND a subject on the backend and we're checking it's been filtered out at some point?

TobyRet replied:
Again my bad - missing an assertion, have fixed. When  a candidate goes to the subjects filter page and selects SEND courses we want to make sure the SEND checkbox is ticked on the results page.

davidgisbey commented on app/views/new_filters/_qualifications_filter.html.erb:
```diff
@@ -0,0 +1,63 @@
+  <div class="govuk-checkboxes govuk-checkboxes--small">
+    <div class="govuk-checkboxes__item">
+      <%=
+        form.check_box(
```
There's probably a good reason, but can we use govuk_check_boxes_fieldset here?

TobyRet replied:
Find does not currently use `gem 'govuk_design_system_formbuilder'`. I tried implementing and ran into some issues.

From PR description  ... "Note that the new filters are not a good candidate for govuk_form_builder, as validations are not required, plus there are some complications around hydrating filter form values from params"

Specifically I struggled to find a way to 'check' the checkbox via params. I think Apply ran into this issue also and are not using `govuk_check_boxes` currently either (where the checked value needs to  be set)

davidgisbey replied:
Yea we have to manually retain the values by messing with values in the controller. Like i said probably a good explanation haha

davidgisbey commented on spec/features/new_results_page_filters/subjects_spec.rb:
```diff
@@ -0,0 +1,329 @@
+      visit subject_path(subjects: %w[31], senCourses: 'true')
+      filter_page.subject_areas.first.subjects[0].checkbox.click # unselect
+      filter_page.subject_areas.first.subjects[1].checkbox.click # select a different one
+      filter_page.send_area.subjects.first.checkbox.click # unselect
```
```
filter_page.subject_areas.first.subjects[0].checkbox.uncheck
filter_page.subject_areas.first.subjects[1].checkbox.check
filter_page.send_area.subjects.first.checkbox.uncheck
```

TobyRet replied:
Copy / paste!
Gone ...
👍

davidgisbey commented on app/views/new_filters/_salary_filter.html.erb:
```diff
@@ -0,0 +1,20 @@
+
+  <div class="govuk-checkboxes govuk-checkboxes--small">
+    <div class="govuk-checkboxes__item" data-qa="subject">
+      <%= form.check_box(
```
can we just use govuk_check_box here?

TobyRet replied:
https://github.com/DFE-Digital/find-teacher-training/pull/650#discussion_r575252853

davidgisbey commented on app/views/new_filters/_send_filter.html.erb:
```diff
@@ -0,0 +1,19 @@
+  </legend>
+  <div class="govuk-checkboxes govuk-checkboxes--small">
+    <div class="govuk-checkboxes__item" data-qa="subject">
+      <%= form.check_box(
```
same here re govuk_check_box

TobyRet replied:
https://github.com/DFE-Digital/find-teacher-training/pull/650#discussion_r575252853

davidgisbey commented on spec/features/new_results_page_filters/study_type_spec.rb:
```diff
@@ -0,0 +1,75 @@
+      end
+    end
+  end
+end
```
is it worth testing that if you deselect both and apply the filter that it returns both values?

TobyRet replied:
Yep good spot - have added a spec

davidgisbey commented on spec/features/new_results_page_filters/vacancies_spec.rb:
```diff
@@ -0,0 +1,51 @@
+
+        expect(results_page.vacancies_filter.subheading.text).to eq('Vacancies')
+        expect(results_page.vacancies_filter.checkbox.checked?).to be(true)
+        expect(results_page.courses.count).to eq(10)
```
I don't know if this is covered somewhere else and it would just be double testing, but should we stub 5 courses with no vacancies and 5 with and check that only 5 are returned when this is set to true?

TobyRet replied:
Yeah more legacy copy and paste.

Given we control the API response via course fixtures there is no added value in  asserting the number of courses displayed in these specs. I've removed

davidgisbey commented on spec/view_objects/result_filters/new_filters_view_spec.rb:
```diff
@@ -0,0 +1,260 @@
+
+        it { is_expected.to eq(true) }
+      end
+
```
it might be overkill, but do you think it's worth adding 

 context 'when across England parameter is present' do
        let(:params) { { l: '2' } }

        it { is_expected.to eq(false) }
      end
end 

 context 'when provider parameter is present' do
        let(:params) { { l: '3' } }

        it { is_expected.to eq(false) }
      end
end 

```

TobyRet replied:
Yep spec are there on lines 127 and 143

https://github.com/DFE-Digital/find-teacher-training/pull/650/files#diff-0c42db30ea8ea34dff0949b189b5d9e7b9c52eb41499587e1666346624bfc958R127

https://github.com/DFE-Digital/find-teacher-training/pull/650/files#diff-0c42db30ea8ea34dff0949b189b5d9e7b9c52eb41499587e1666346624bfc958R143

davidgisbey replied:
I think they're testing a different method though right? At the moment we only test that it returns false when it's blank and not that it only returns true if   l == 1. It's really not a big deal anyway though 👍

davidgisbey commented on app/views/results/index.html.erb:
```diff
@@ -1,11 +1,6 @@
-<% if FeatureFlag.active?(:new_filters) %>
-  <h1 class="govuk-heading-l" data-qa="heading">
-    <span class="govuk-caption-l">Teacher training courses</span>
-    <%= "#{@results_view.number_of_courses_string} found" %>
```
Do we not need this then?

davidgisbey commented:
Looking great Toby 👍 There's one bug and a bunch of nits, but it's looking really excellent

davidgisbey commented on app/views/new_filters/_location_provider_filter.html.erb:
```diff
@@ -0,0 +1,30 @@
+      <%= request.parameters['lq'] %>
+    <% elsif @filters_view.provider_query? %>
+      <%= request.parameters['query'] %>
+    <% end %>
```
I think we need to pass these values through as hidden fields 

    <%= form.hidden_field(:query, value: params[:query]) %>
      <%= form.hidden_field(:lat, value: params[:lat]) %>
      <%= form.hidden_field(:lng, value: params[:lng]) %>
      <%= form.hidden_field(:loc, value: params[:loc]) %>
      <%= form.hidden_field(:lq, value: params[:lq]) %>

TobyRet replied:
Yep - that has been done now

davidgisbey reviewed (approved):
Great work 👌
